third_party/terraform/utils/batcher.go

```diff
@@ -0,0 +1,204 @@
+	CombineF batcherCombineFunc
+
+	// SendF function determines how to actually send a batched request to a
+	// third party service. The arguments given to this function are
+	// (ResourceName, Body) where Body may have been combined with other request
+	// Bodies.
+	SendF batcherSendFunc
+
+	// ID for debugging request. This should be specific to a single request
+	// (i.e. per Terraform resource)
+	DebugId string
+}
+
+// These types are meant to be the public interface to batchers. They define
+// logic to manage batch data type and behavior, and require service-specific
+// implementations per type of request per service.
+// Function type for combine existing batches and additional batch data
+type batcherCombineFunc func(body interface{}, toAdd interface{}) (interface{}, error)
+
+// Function type for sending a batch request
+type batcherSendFunc func(resourceName string, body interface{}) (interface{}, error)
+
+// batchResponse bundles an API response (data, error) tuple.
+type batchResponse struct {
+	body interface{}
+	err  error
+}
+
+// startedBatch refers to a processed batch whose timer to send the request has
+// already been started. The responses for the request is sent to each listener
+// channel, representing parallel callers that are waiting on requests
+// combined into this batch.
+type startedBatch struct {
+	*BatchRequest
+
+	listeners []chan batchResponse
+	timer     *time.Timer
+}
+
+// batchingConfig contains user configuration for controlling batch requests.
+type batchingConfig struct {
+	sendAfter       time.Duration
+	disableBatching bool
+}
+
+// Initializes a new batcher
+func NewRequestBatcher(config *batchingConfig) *RequestBatcher {
+	return &RequestBatcher{
+		batchingConfig: config,
+		batches:        make(map[string]*startedBatch),
+	}
+}
+
+// SendRequestWithTimeout is expected to be called per parallel call.
+// It manages waiting on the result of a batch request.
+//
+// Batch requests are grouped by the given batchKey, and thus batchKey
+// should be unique to the type of batch request being sent and the affected
+// third-party resource.
+func (b *RequestBatcher) SendRequestWithTimeout(batchKey string, request *BatchRequest, timeout time.Duration) (interface{}, error) {
+	if request == nil {
+		return nil, fmt.Errorf("error, cannot request batching for nil BatchRequest")
+	}
+	if request.CombineF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil CombineF")
+	}
+	if request.SendF == nil {
+		return nil, fmt.Errorf("error, cannot request batching for BatchRequest with nil SendF")
+	}
+	if b.disableBatching {
+		log.Printf("[DEBUG] Batching is disabled, sending single request for %q", request.DebugId)
+		return request.SendF(batchKey, request.Body)
+	}
+
+	respCh, err := b.startBatchRequest(batchKey, request)
+	if err != nil {
+		return nil, fmt.Errorf("error adding request to batch: %s", err)
+	}
+
+	timer := time.NewTimer(timeout)
```

Comment:
Seems like this should be a cancellable context WithTimeout instead of using your own timer. This will allow upstream cancellations (such as UpdateTimeout) to propagate downwards.

Reply:
By upstream I mean upstream in the code (not requests). If somebody ctl-c's the terraform process and kills the binary or if the UpdateTimeout is reached, it would be nice to gracefully cancel the in flight batches. I think you can pass the `context` from config into the batcher and then wrap it with [a `WithTimeout`](https://golang.org/pkg/context/#example_WithTimeout). We don't do a lot of context propagation in the provider right now but I think it will be valuable if we want to get better at graceful termination in general.

Reply:
ah true I wasn't thinking about using the context from Config